### ISSUE TITLE
[suggestion] monster onThink improvements

### ISSUE THREAD

**Issue by Zbizu** [otland/forgottenserver#3003]
What if we moved refreshing monster aggro from spawns to players and summons?

In case the question wasn't clear enough, I'm speaking of spawn.cpp loop that calls onCreatureAppear.

### current problem
if there are 10000+ monsters, every single spawn calls getSpectators. RL maps and custom rpg maps tend to have that much, even high exp servers do, though they have less spawns. Tracked maps have one monster per spawn which can also get laggy.

### potential solution
biggest servers have around 2000 players. If we moved aggro checks (onCreatureAppear) to player and his summons and label monsters that they were just checked (to avoid duplicate checks), we could reduce amount of getSpectators calls (though the calls would return more data because players are usually together in cities).

I don't know which one would be faster, but this solution should definitely be explored.

### difficulty
high (too high for me) - it needs to be completely rewritten and needs a lot of testing and performance checks

**Comment by MillhioreBT** [comment]
You could do a bitmap (0/1) that updates when the player changes his position and using that map look for the status (1) in the range with a small loop and thus know if a player is in field of vision, completely eliminating the use of the function **getSpectators**, I don't know if this consumes more time, although in theory it shouldn't, since it only handles a map of zeros and ones

**Comment by Zbizu** [comment]
current monster activity structure:
checkSpawn -> loop through all spawns on the server (!) -> check for spawn block -> spawn monster -> add it to checkCreatures vector -> loop through all creatures on the server (!!)

suggested change:
load spawns -> load monsters -> if the monster has onThink, add him to separate table to loop through onThink events -> loop through (1000 if we're being optimistic) online players to call getSpectators (instead of ~60k monsters/spawns calling it) -> add spawns/monsters triggered by player to list of "disrupted" things(use a new field "lastUpdated" with timestamp to make sure nothing gets called twice) -> keep updating disrupted things until they come back to normal (monsters walked back after few seconds of being alone or they all respawned and spawn has nothing to do)

tl;dr: instead of every monster checking what's around, players would trigger them and with this change only disrupted things would be looped through
